### ISSUE TITLE
Firefox does not correctly keep focus on parent window.  Instead the focus changes to the child window incorrectly.

### ISSUE THREAD

**Issue by iamle0pard** [DevExpress/testcafe#5797]
### What is your Test Scenario?
....our scenario is a "chat application" that allows you to type messages between users in two different browsers...

Start your test using one window that has a text input field.  This will be Browser 1
* In our case when you put "focus" in this input field, the input border changes color (to a blue border)
Spawn a child window that also has a text input field.  This will be Browser 2
Use one of the testcafe APIs to switch back to the parent window (Browser 1) we have used "switchToWindow"
Type some text into the input field in Browser 1 and hit "Enter" to send it
* Remember, setting focus to this input field will get the CSS style to change
Note: Since this is a chat application, text sent in Browser 1 will show up in Browser 2.  When done using Firefox the child browser (browser 2) will now have focus
In your test, try to verify the CSS style of the input field on Browser 1
This will fail because the focus is currently on Browser 2 (incorrectly)
 
### What is the Current behavior?
Firefox incorrectly changes focus to the child window and does not apply the CSS styles to the parent window (when the parent window is visible in the background).  Chrome handles this just fine.
 
### What is the Expected behavior?
We expect that when we use "switchToWindow" or similar method that the focus stays in that specific window, even if it is visibly in the background.

Alternatively we would like a way to "bring window to foreground" to prevent this behavior from happening.  We tried to "sendKeys('alt+tab')" but that didn't work.  also doing a "t.click('input')" on the input box doesn't bring that browser visibly to the foreground (when viewed in a non-headless run).
 
### What is your web application and your TestCafe test code?
This is an internal chat application for our company that requires credentials to access.  This is why I tried to make the test scenario generic to say "this is a chat application" so that you could understand the mechanics of sending messages between the two browsers without needing our specific code.

### Sample Code:
```
import { Selector } from 'testcafe';

let childWindow;

fixture('Sample test to reproduce issues...')
    .page('https://www.yahoo.com')
    .beforeEach(async (t) => {
        childWindow = await t.openWindow('https://www.google.com');
        await t.switchToParentWindow();
    });

test('With Firefox :focus while child window is open.', async (t) => {
    const searchInput = Selector('.ybar-ytheme-fuji2 ._yb_cerck'); // yahoo search box
    await t.expect(searchInput.focused).eql(true);
    await t
        .expect(searchInput.getStyleProperty('box-shadow'))
        .eql('rgba(0, 0, 0, 0.075) 0px 1px 1px 0px inset, rgba(82, 168, 236, 0.6) 0px 0px 8px 0px');
});
```
Note: Please test this using Firefox.  If you use this sample code to reproduce, you will see the test only passes if you remove the code within the 'beforeEach' (comment it out or remove).  We expect the test to pass in both scenarios.

### Steps to Reproduce:
1. Launch a browser as a chat user #1 
2. Launch a child browser as chat user #2
3. SwitchToWindow (parent window in #1)
4. Type some text in input field, verify CSS styles are set
5. Hit Enter to send this text
Note: when you hit Enter the application will bring focus right back to input field so you can continue typing/chatting
5. Verify the CSS of the input field (still in user #1 browser)
6. Notice that the focus is actually in user #2 browser (this only happens in Firefox.. chrome handles this fine)
 
### Your Environment details:
 
* testcafe version:                   1.9.4
* node.js version:                    10.15.0
* command-line arguments:  
* browser name and version: Chrome: Version 87.0.4280.88 (Official Build) (x86_64), Firefox: 83.0
* platform and version:          Windows 10, MacOS Catalina Version 10.15.7

**Comment by SergeyShurygin** [comment]
Hi,

I investigated the issue and found the following:

The Yahoo search box's input element in the parent page has a correct focused attribute after switching windows and the test successfully passes this line `await t.expect(searchInput.focused).eql(true) `in both browsers.

In Chrome, the test fails on the last assert statement when checking the box-shadow style of the input. Firefox works fine for me. As far as I understand, Chrome will not apply a focused style on the input until its window becomes the foreground. If you run the test with the --debug-on-fail option and the test fails, Yahoo's search box receives a shadow border and shows caret only when you click in the browser's window and make its window the foreground.

We did not design the multiple windows feature to bring windows in the foreground when switching them. Thus, this behavior can be considered as a feature, not a bug.

I can suggest two options to test your scenario.

The first is to check the input's focused attribute only and let the browser apply correct styles by itself (as it should). You can perform a separate test that ensures that the input has the correct style on the first page load. At the same time, perform another test checking only the focused attribute after switching windows in a multiple pages scenario. Our team considers this way to be preferable.

The second way is to use the bringWindowToFront function from testcafe-browser-tools package after switching windows. Here is the sample code (I changed the Yahoo search box selector because the selector from your sample didn't work for me):

```
import { Selector } from 'testcafe';
import { ClientFunction } from 'testcafe';
import { bringWindowToFront } from 'testcafe-browser-tools';

let childWindow;

fixture('Sample test to reproduce issues...')
    .page('https://www.yahoo.com')
    .beforeEach(async (t) => {
        const getMainWindowTitle = ClientFunction(() => window['%hammerhead%'].nativeMethods.documentTitleGetter.call(document));
        t.ctx.mainWindowTitle = await getMainWindowTitle();

        childWindow = await t.openWindow('https://www.google.com');
        await t.switchToParentWindow();    
    });

test('With Firefox :focus while child window is open.', async (t) => {
    const searchInput = Selector('#ybar-sbq'); // yahoo search box
    bringWindowToFront(t.ctx.mainWindowTitle);
    await t.debug();
    await t.expect(searchInput.focused).eql(true);
    await t
        .expect(searchInput.getStyleProperty('box-shadow'))
        .eql('rgba(0, 0, 0, 0.075) 0px 1px 1px 0px inset, rgba(82, 168, 236, 0.6) 0px 0px 8px 0px');
});
```

**Comment by iamle0pard** [comment]
Thank you so much for the detailed response and suggestions!  

Unfortunately the first option doesn't work for us as the textarea we're testing only appears in a 2 browser scenario because it doesn't make sense to have a chat input without a chat.

The second option only worked for us in non-headless mode (with the browser UI visible) and we regularly do our testing in headless mode (as part of a build pipeline).  We did some additional logging and noticed that even in headless mode there is a 'mainWindowTitle' to refer to, but the test was failing which indicated to us that it's not actually in the foreground.

Do you know if there is something that would work in headless mode (or why headless mode would give different results than non-headless)?

I do want to mention that I agree that the .focus attribute was true in that only the style is not applied, but the point of our test is to do a CSS style check.  Do you consider testing styles this way bad practice or anything like that?  We are trying to do it to ensure the look and feel of our UI don't change over time.

**Comment by SergeyShurygin** [comment]
I can't reproduce the issue with Chrome Headless. My sample test works fine in headless mode too. This may be an OS-specific problem. We need a working sample and information about the OS and the browser versions to give a precise answer.

>Do you consider testing styles this way bad practice or anything like that?

It is definitely not bad practice in a general case. For a specific scenario, if you want to test the behavior of your application, e.g., that the input receives focus on page load, we consider it excessive. However, if you want to do some kind of visual regressive testing to ensure that the control's visual appearance doesn't change over time - it makes sense. In this case, it is important not to combine different test scenarios - behavior and visual testing in a single test.

**Comment by iamle0pard** [comment]
@SergeySeleznyov - here is our exact code that we are using to reproduce this issue.

**runner.ts:**
----------
```typescript
import createTestCafe from 'testcafe';

let testcafe;

const browsers = ['chrome:headless', 'firefox:headless'];

async function runTestForBrowser(browser: string) {
    try {
        // Start running E2E tests
        testcafe = await createTestCafe('localhost');
        const runner = testcafe.createRunner();

        const testSource = ['./test/e2e/chat/sample.e2e.ts'];

        // Log the current bacUrl to assist with debugging failed tests
        // console.log(`process.env.bacUrl: ${process.env.bacUrl}`);

        const failedCount = await runner
            .src(testSource)
            .browsers(browser)
            .reporter([
                'spec', // Output to STDOUT (console)
                {
                    name: 'xunit',
                    output: 'test/reports/bacUIe2e.xml',
                },
                {
                    name: 'html',
                    output: 'test/reports/bacUIe2e.html',
                },
            ])
            .screenshots({
                path: 'test/reports/screenshots/',
                takeOnFails: true,
                pathPattern: '${DATE}_${TIME}/test-${TEST_INDEX}/${USERAGENT}/${FILE_INDEX}.png',
            })
            .run({
                debugMode: false,
                debugOnFail: true,
                stopOnFirstFail: false,
                skipJsErrors: true,
                skipUncaughtErrors: true,
                selectorTimeout: 30000,
                assertionTimeout: 10000,
                pageLoadTimeout: 1000,
                speed: 1,
                disablePageCaching: false,
                developmentMode: false,
            })
            .catch((error) => {
                console.log(error);
                throw new Error(error);
            });

        if (failedCount > 0) {
            console.log('Tests failed: ' + failedCount);
            if (testcafe) {
                testcafe.close();
            }
            throw new Error(`TestCafe failed to run. Failed tests: ${failedCount}`);
        }

        return;
    } catch (error) {
        if (testcafe) {
            testcafe.close();
        }
        throw new Error(`TestCafe failed to run. ${error}`);
    }
}

const runAllBrowsers = async () => {
    for (const browser of browsers) {
        await runTestForBrowser(browser);
    }
    process.exit();
};

runAllBrowsers();
```


**sample.e2e.ts:**
-------------------
```typescript
import { Selector, ClientFunction } from 'testcafe';
import { bringWindowToFront } from 'testcafe-browser-tools';

fixture('Sample test to reproduce issues...')
    .page('https://www.yahoo.com')
    .beforeEach(async (t) => {
        const getMainWindowTitle = ClientFunction(() =>
            window['%hammerhead%'].nativeMethods.documentTitleGetter.call(document),
        );
        t.ctx.mainWindowTitle = await getMainWindowTitle();
        await t.openWindow('https://www.google.com');
        await t.switchToParentWindow();
    });

test('With Firefox :focus while child window is open.', async (t) => {
    bringWindowToFront(t.ctx.mainWindowTitle);
    const searchInput = Selector('#ybar-sbq'); // yahoo search box
    await t.expect(searchInput.focused).eql(true);
    await t
        .expect(searchInput.getStyleProperty('box-shadow'))
        .eql('rgba(0, 0, 0, 0.075) 0px 1px 1px 0px inset, rgba(82, 168, 236, 0.6) 0px 0px 8px 0px');
});
```

**tsconfig.ts:**
----------------
```typescript
{
  "compilerOptions": {
    "allowSyntheticDefaultImports": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "preserveConstEnums": false,
    "moduleResolution": "node",
    "jsx": "react",
    "lib": ["dom", "es2019"],
    "target": "es2019",
    "skipLibCheck": true
  },
  "exclude": [
    "node_modules",
    "static"
  ]
}
```

Here are the configurations that we have seen this issue reproduce on:

**Windows 10 Home**
Firefox: 84.0.1 (64-bit)
Chrome: 87.0.4280.88 (Official Build) (64-bit)
Screenshot of failure: https://i.imgur.com/21KNHnf.png

**Mac OS Catalina**
Version: 10.15.7
Firefox: 84.0.1 (64-bit)
Chrome: 87.0.4280.88 (Official Build) (x86_64)
Screenshot of failure: https://i.imgur.com/4cVUBWj.png

**Comment by AlexKamaev** [comment]
I checked the suggested solution in Chrome: headless and Firefox: headless and found that for me it's working in headless Chrome, but it's still not working in headless FF.
I suppose that we cannot use the `testcafe-browser-tools` functions in headless mode.
I can suggest another workaround for FF. Please note that it works only in the latest v1.10.1. In addition, I want to note that the workaround uses an internal API that can be changed in the future, so you need to use it at your own risk.

```JS
import { Selector, ClientFunction } from 'testcafe';

fixture('Sample test to reproduce issues...')
    .page('https://www.yahoo.com')
    .beforeEach(async (t) => {
        const getMainWindowTitle = ClientFunction(() =>
            window['%hammerhead%'].nativeMethods.documentTitleGetter.call(document),
        );
        t.ctx.mainWindowTitle = await getMainWindowTitle();
        await t.openWindow('https://www.google.com');
        await t.switchToParentWindow();
    });

test('With Firefox :focus while child window is open.', async (t) => {
    const browserConnection = t.testRun.browserConnection;
    const marionetteClient  = browserConnection.provider.plugin.openedBrowsers[browserConnection.id].marionetteClient;

    if (marionetteClient && marionetteClient._ensureActiveWindow)
        await marionetteClient._ensureActiveWindow();

    const searchInput = Selector('#ybar-sbq'); // yahoo search box
    await t.expect(searchInput.focused).eql(true);
    await t
        .expect(searchInput.getStyleProperty('box-shadow'))
        .eql('rgba(0, 0, 0, 0.075) 0px 1px 1px 0px inset, rgba(82, 168, 236, 0.6) 0px 0px 8px 0px');
});
```
The idea is to call the `marionetteClient._ensureActiveWindow();` method. Please note that in Chrome you still need to use the `bringWindowToFront` workaround.

**Comment by iamle0pard** [comment]
Thanks for this suggestion and sample code, we did attempt to use it and updated our TestCafe to 1.10.1 however we noticed a new bug that seems to be showing up only with 1.10.1 in openWindow. It seems to be loading the page twice in rapid succession, which in most software might not be noticable but our system restricts the same person from visiting the page to every 30 seconds.  I will be filing a separate issue about this today.  I believe we can work-around this until it's fixed, but wanted to call it out here.

Other than this bug, the sample code you provided does seem to help us with getting our test to pass.
